rfcs/0008-storage-handler.md

```diff
@@ -0,0 +1,338 @@
+
+> **Note**
+> Struct variables are laid out in the same way today.
+
+For example, the storage variable `str` above is a string containing `99` characters and its size is `99` bytes which span 2 storage slots. It is also assigned index `2`. Therefore, `str` spans slots with keys `sha256("storage_2")` and `sha256("storage_2") + 1`.
+
+Variable `s` on the other hand, is a nested struct and is assigned index `3`. Therefore, each of its fields (and subfields) is assigned a separate key starting with `sha256("storage_3")`:
+
+- `s.a` is assigned `sha256("storage_3")`.
+- `s.b.c` is assigned `sha256("storage_3") + 1`.
+- `s.b.d` is assigned `sha256("storage_3") + 2`.
+
+### Empty Types
+
+There are two possible empty types: structs with no fields and enums with no variants. Empty structs are useful for implementing dynamic storage types such as `StorageMap` and `StorageVec`. Storage variables that have empty types should not consume any storage slots.
+
+### Behavior of `StorageKey`
+
+When a storage variable is accessed, a `StorageKey` is returned of the appropriate type. The `StorageKey` object contains the `b256` key required to access the underlying data:
+
+```rust
+assert(storage.x.key == sha256("storage_0"));
+assert(storage.y.key == sha256("storage_1"));
+assert(storage.str.key == sha256("storage_2"));
+assert(storage.s.a.key == sha256("storage_3"));
+assert(storage.s.b.c.key == sha256("storage_3") + 1);
+assert(storage.s.b.d.key == sha256("storage_3") + 2);
+
+assert(storage.s.key == sha256("storage_3")); // Same as `storage.s.a.key`
+assert(storage.s.b.key == sha256("storage_3") + 1); // Same as `storage.s.b.c.key`
+```
+
+> **Note**: 
+> The above is a pseudo-code and is not meant to compile for various reasons.
+
+## Implementation Detail
+
+The typed expression `TyStorageAccess` in the compiler should now return `std::storage::StorageKey` which should store the key chosen by the compiler according to the rules described in [Storage Slots Assignment](#storage-slot-assignment). The rest of the flow is be handled automatically after removing all the unnecessary logic in the compiler for reading and writing storage slots.
+
+## Implementing Dynamic Storage Types using `StorageKey`.
+
+Dynamic storage types such `StorageMap` and `StorageVec` currently use the intrinsic `__get_storage_key` which is not very well defined and has a fragile implementation. With `StorageKey` available and returned directly from a storage access expression (as in `storage.my_map`), we can re-implement methods like `StorageMap::insert`, `StorageMap::get`, and `StorageMap::remove` in a safer and cleaner way as follows:
```

Comment:
I know very little about `__get_storage_key ` and why it's bad. It would be helpful if there was a section going into a little more detail about why it's important to remove the usage of `__get_storage_key`, since it sounds like a key motivator for this RFC.

I'm wondering:
1. What does it mean to say an intrinsic is not well defined?
1. What is fragile about the implementation?
1. What are the risks or vulnerabilities with using it?
1. Why was it used in the first place?

Reply:
I will make the motivation section clearer!

Reply:
I added a sentence about why its implementation is bad. This is mostly the big reason for removing this. The only way to implement it is bad because it relies on inlining.